File: installer/src/main/java/com/teamscale/profiler/installer/steps/InstallSystemdStep.java
package com.teamscale.profiler.installer.steps;

import com.teamscale.profiler.installer.EnvironmentMap;
import com.teamscale.profiler.installer.FatalInstallerError;
import com.teamscale.profiler.installer.PermissionError;
import com.teamscale.profiler.installer.TeamscaleCredentials;
import org.conqat.lib.commons.system.SystemUtils;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.TimeUnit;

/**
 * On Linux with systemd, registers the agent globally for systemd services. This is necessary in addition to
 * {@link InstallEtcEnvironmentStep}, since systemd doesn't always inject /etc/environment into started services.
 */
public class InstallSystemdStep implements IStep {

	private final Path etcDirectory;
	private final EnvironmentMap environmentVariables;
	private final boolean reloadSystemdDaemon;

	public InstallSystemdStep(Path etcDirectory, EnvironmentMap environmentMap, boolean reloadSystemdDaemon) {
		this.etcDirectory = etcDirectory;
		this.environmentVariables = environmentMap;
		this.reloadSystemdDaemon = reloadSystemdDaemon;
	}

	@Override
	public void install(TeamscaleCredentials credentials) throws FatalInstallerError {
		if (!SystemUtils.isLinux()) {
			return;
		}

		if (!Files.exists(getSystemdEtcDirectory())) {
			System.out.println("systemd could not be detected. Not installing profiler for systemd services.");
			// system has no systemd installed
			return;
		}

		if (!Files.exists(getSystemdSystemConfDDirectory())) {
			try {
				Files.createDirectories(getSystemdSystemConfDDirectory());
			} catch (IOException e) {
				throw new PermissionError("Cannot create system.conf.d directory: " + getSystemdSystemConfDDirectory(),
						e);
			}
		}

		Path systemdConfigFile = getSystemdConfigFile();
		if (Files.exists(systemdConfigFile)) {
			throw new PermissionError(
					"Cannot create systemd configuration file " + systemdConfigFile + " because it already exists." +
							"\nPlease uninstall any old profiler versions first");
		}

		String content = "[Manager]\nDefaultEnvironment=" + environmentVariables.getSystemdString() + "\n";
		try {
			Files.write(systemdConfigFile, content.getBytes(StandardCharsets.UTF_8));
		} catch (IOException e) {
			throw new PermissionError("Could not create " + systemdConfigFile, e);
		}

		daemonReload();
	}

	private void daemonReload() {
		if (!this.reloadSystemdDaemon) {
			return;
		}

		try {
			ProcessBuilder builder = new ProcessBuilder("systemctl", "daemon-reload")
					// must redirect program output or it might hang forever
					.redirectError(ProcessBuilder.Redirect.to(new File("/dev/null")))
					.redirectOutput(ProcessBuilder.Redirect.to(new File("/dev/null")));

			Process process = builder.start();
			if (!process.waitFor(5, TimeUnit.SECONDS) || process.exitValue() != 0) {
				// timeout
				process.destroyForcibly();
				askUserToManuallyReloadDaemon();
			}
		} catch (IOException | InterruptedException e) {
			e.printStackTrace();
			askUserToManuallyReloadDaemon();
		}
	}

	private void askUserToManuallyReloadDaemon() {
		System.err.println(
				"Failed to reload the systemd daemon. Systemd services can only be profiled after reloading the daemon." +
						"\nPlease manually reload the daemon with:" +
						"\nsystemctl daemon-reload");
	}

	private Path getSystemdEtcDirectory() {
		return etcDirectory.resolve("systemd");
	}

	private Path getSystemdSystemConfDDirectory() {
		return getSystemdEtcDirectory().resolve("system.conf.d");
	}

	private Path getSystemdConfigFile() {
		return getSystemdSystemConfDDirectory().resolve("teamscale-java-profiler.conf");
	}

	@Override
	public void uninstall(IUninstallErrorReporter errorReporter) {
		if (!SystemUtils.isLinux()) {
			return;
		}

		Path systemdConfigFile = getSystemdConfigFile();
		if (!Files.exists(systemdConfigFile)) {
			return;
		}

		try {
			Files.delete(systemdConfigFile);
		} catch (IOException e) {
			errorReporter.report(
					new PermissionError("Failed to remove systemd config file " + systemdConfigFile + "." +
							" Manually remove this file or systemd Java services may fail to start.", e));
		}

		daemonReload();
	}
}

File: installer/src/main/kotlin/com/teamscale/profiler/installer/steps/InstallSystemdStep.kt
package com.teamscale.profiler.installer.steps

import com.teamscale.profiler.installer.*
import com.teamscale.profiler.installer.steps.IStep.IUninstallErrorReporter
import org.apache.commons.lang3.SystemUtils
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.concurrent.TimeUnit

/**
 * On Linux with systemd, registers the agent globally for systemd services. This is necessary in addition to
 * [InstallEtcEnvironmentStep], since systemd doesn't always inject /etc/environment into started services.
 */
class InstallSystemdStep(
	private val etcDirectory: Path,
	private val environmentVariables: JvmEnvironmentMap,
	private val reloadSystemdDaemon: Boolean
) : IStep {
	override fun shouldRun() = SystemUtils.IS_OS_LINUX

	@Throws(FatalInstallerError::class)
	override fun install(credentials: TeamscaleCredentials) {
		if (!Files.exists(systemdEtcDirectory)) {
			println("systemd could not be detected. Not installing profiler for systemd services.")
			// system has no systemd installed
			return
		}

		if (!Files.exists(systemdSystemConfDDirectory)) {
			try {
				Files.createDirectories(systemdSystemConfDDirectory)
			} catch (e: IOException) {
				throw PermissionError("Cannot create system.conf.d directory: $systemdSystemConfDDirectory", e)
			}
		}

		val systemdConfigFile = systemdConfigFile
		if (Files.exists(systemdConfigFile)) {
			throw PermissionError(
				"""
				Cannot create systemd configuration file $systemdConfigFile because it already exists.
				Please uninstall any old profiler versions first
				""".trimIndent()
			)
		}

		val content = """
			[Manager]
			DefaultEnvironment=${environmentVariables.systemdString}
			""".trimIndent()
		try {
			Files.writeString(systemdConfigFile, content)
		} catch (e: IOException) {
			throw PermissionError("Could not create $systemdConfigFile", e)
		}

		daemonReload()
	}

	private fun daemonReload() {
		if (!this.reloadSystemdDaemon) {
			return
		}

		try {
			val builder =
				ProcessBuilder("systemctl", "daemon-reload") // must redirect program output or it might hang forever
					.redirectError(ProcessBuilder.Redirect.to(File("/dev/null")))
					.redirectOutput(ProcessBuilder.Redirect.to(File("/dev/null")))

			val process = builder.start()
			if (!process.waitFor(5, TimeUnit.SECONDS) || process.exitValue() != 0) {
				// timeout
				process.destroyForcibly()
				askUserToManuallyReloadDaemon()
			}
		} catch (e: IOException) {
			e.printStackTrace()
			askUserToManuallyReloadDaemon()
		} catch (e: InterruptedException) {
			e.printStackTrace()
			askUserToManuallyReloadDaemon()
		}
	}

	private fun askUserToManuallyReloadDaemon() {
		System.err.println(
			"""
			Failed to reload the systemd daemon. Systemd services can only be profiled after reloading the daemon.
			Please manually reload the daemon with:
			systemctl daemon-reload
			""".trimIndent()
		)
	}

	private val systemdEtcDirectory: Path
		get() = etcDirectory.resolve("systemd")

	private val systemdSystemConfDDirectory: Path
		get() = systemdEtcDirectory.resolve("system.conf.d")

	private val systemdConfigFile: Path
		get() = systemdSystemConfDDirectory.resolve("teamscale-java-profiler.conf")

	override fun uninstall(errorReporter: IUninstallErrorReporter) {
		val systemdConfigFile = systemdConfigFile
		if (!Files.exists(systemdConfigFile)) {
			return
		}

		try {
			Files.delete(systemdConfigFile)
		} catch (e: IOException) {
			errorReporter.report(
				PermissionError("Failed to remove systemd config file $systemdConfigFile. Manually remove this file or systemd Java services may fail to start.", e)
			)
		}

		daemonReload()
	}
}
